File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(root==NULL){
            return 0;
        }   
        int dl = diameterOfBinaryTree(root->left);
        int dr = diameterOfBinaryTree(root->right);
        int curr = heightOfTree(root->left) + heightOfTree(root->right);
        return max(curr, max(dl, dr));
    }
    int heightOfTree(TreeNode* root){
        if(root==NULL){
            return 0;
        }
        return 1+max(heightOfTree(root->left), heightOfTree(root->right));
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int diameterOfBinaryTree(TreeNode root) {
        if(root==null){
            return 0;
        }   
        int dl = diameterOfBinaryTree(root.left);
        int dr = diameterOfBinaryTree(root.right);
        int curr = heightOfTree(root.left) + heightOfTree(root.right);
        return Math.max(curr, Math.max(dl, dr));
    }
    int heightOfTree(TreeNode root){
        if(root==null){
            return 0;
        }
        return 1+Math.max(heightOfTree(root.left), heightOfTree(root.right));
    }
}